pretyping/inferCumulativity.ml

```diff
@@ -12,41 +12,52 @@ open Reduction
 open Declarations
 open Constr
 open Univ
+open Variance
 open Util
 
 (** Throughout this module we modify a map [variances] from local
-   universes to [Variance.t]. It starts as a trivial mapping to
-   [Irrelevant] and every time we encounter a local universe we
-   restrict it accordingly. *)
+    universes to [Variance.t]. It starts as a trivial mapping to
+    [Irrelevant] and every time we encounter a local universe we
+    restrict it accordingly. If we ever detect that all universes are
+    [Invariant] we raise [TrivialVariance] to avoid doing extra work.
+*)
+
+type variances = {
+  v : Variance.t LMap.t;
+  nontrivial_count : int;
+}
+exception TrivialVariance
+
+let maybe_trivial variances =
+  if variances.nontrivial_count = 0 then raise TrivialVariance
+  else variances
 
 let infer_level_eq u variances =
-  if LMap.mem u variances
-  then LMap.set u Variance.Invariant variances
-  else variances
+  match LMap.set u Variance.Invariant variances.v with
```

Comment:
This looks pretty fragile to me. If ever we break the invariant that adding the same data in a map results in physical equality, this code is broken. I'd rather check the current value, even if that incurs one more access, or use the `modify` function.